internal/credential/store.go

```diff
@@ -0,0 +1,114 @@
+package credential
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/docker/cli/cli/config"
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/docker/cli/cli/config/credentials"
+	"github.com/docker/cli/cli/config/types"
+	"oras.land/oras-go/v2/registry/remote/auth"
+)
+
+// Store provides credential CRUD operations.
+type Store struct {
+	configs []*configfile.ConfigFile
+}
+
+// NewStore generates a store based on the passed in config file path.
+func NewStore(configPaths ...string) (*Store, error) {
+	if len(configPaths) == 0 {
+		// No config path passed, load default docker config file.
+		cfg, err := config.Load(config.Dir())
+		if err != nil {
+			return nil, err
+		}
+		if !cfg.ContainsAuth() {
+			cfg.CredentialsStore = credentials.DetectDefaultStore(cfg.CredentialsStore)
+		}
+
+		return &Store{
+			configs: []*configfile.ConfigFile{cfg},
+		}, nil
+	}
+
+	var configs []*configfile.ConfigFile
+	for _, path := range configPaths {
+		cfg, err := loadConfigFile(path)
+		if err != nil {
+			return nil, fmt.Errorf(path, err)
+		}
+		configs = append(configs, cfg)
+	}
+
+	return &Store{
+		configs: configs,
+	}, nil
+}
+
+// loadConfigFile reads the configuration files from the given path.
```

Comment:
Sorry I still don't get this. It reads configuration _files_ from the given path? Can we clarify what do we do if the path does not exist?

Reply:
1) It only reads the configuration
- If the config file exists, then read from file
- If not, then create a in-memory object, which will be persistent later when store is called

2) why the code here is hard to understand
The [LoadFromReader](https://pkg.go.dev/github.com/docker/cli@v20.10.14+incompatible/cli/config/configfile#ConfigFile.LoadFromReader) function doesn't have a static counter part, so a new config always need to be initalized first, which make the if-else structure weird. I have re-structured the code, although the `cfg = configfile.New(path)` is called twice, the code is more understandable. Please check